source/client/gui/screens/ControlsScreen.cpp

```diff
@@ -0,0 +1,245 @@
+
+		flipLeft = !flipLeft;
+		m_rebindingOptions[i] = newOption;
+	}
+
+}
+
+ControlsScreen::~ControlsScreen()
+{
+	for (int i = 0; i < KM_COUNT; i++)
+	{
+		delete m_rebindingOptions[i].button;
+	}
+}
+
+void ControlsScreen::UpdateText()
+{
+	Options& o = m_pMinecraft->m_options;
+
+	for (int i = 0; i < KM_COUNT; i++)
+	{
+		Button* button = m_rebindingOptions[i].button;
+		button->m_text = GetCharacterText(button->m_buttonId);
+	}
+}
+
+void ControlsScreen::GetRenderableButtons()
+{
+	int firstRenderable = (m_iScrollHeight / BUTTON_GAP) * 2;
+	int visibleArea = (m_height - TOP_CUTOFF - BOTTOM_CUTOFF - 15);
+	int lastRenderable = firstRenderable + ((visibleArea + BUTTON_GAP ) / BUTTON_GAP) * 2;
+
+	if (lastRenderable >= KM_COUNT)
+		lastRenderable = KM_COUNT;
+	for (int i = 0; i < KM_COUNT; i++)
+	{
+		if (i >= firstRenderable && i < lastRenderable)
+		{
+			m_rebindingOptions[i].button->m_bEnabled = true;
+			continue;
+		}
+		m_rebindingOptions[i].button->m_bEnabled = false;
+	}
+
+	m_iFirstRenderable = firstRenderable;
+	m_iLastRenderable = lastRenderable;
+}
+
+std::string ControlsScreen::GetCharacterText(int keycode)
+{
+	Options& o = m_pMinecraft->m_options;
+
+	int keyCode = o.m_keyMappings[keycode].value;
+	return m_pMinecraft->m_pPlatform->convertKeycodeToString(keyCode);
+}
+
+void ControlsScreen::mouseEvent()
+{
+	MouseAction* pAction = Mouse::getEvent();
+	if (pAction->_buttonType == Mouse::ButtonType::SCROLLWHEEL)
+	{
+
+		if (Mouse::getEventButtonState() <= 0) // @NOTE: Scroll up
+		{
+			m_iScrollHeight -= BUTTON_GAP;
+			if (m_iScrollHeight <= 0)
+			{
+				m_iScrollHeight = 0;
+			}
+		}
+		else
+		{
+			m_iScrollHeight += BUTTON_GAP;
+			if (m_iScrollHeight > m_iHighestScrollHeight)
+			{
+				m_iScrollHeight = m_iHighestScrollHeight;
+			}
+		}
+	}
+
+	GetRenderableButtons();
+	Screen::mouseEvent();
+}
+
+void ControlsScreen::init()
+{ 
+	m_iHighestScrollHeight = (BUTTON_GAP * ((int)round(KM_COUNT / 2.0f) - ((m_height - TOP_CUTOFF - BOTTOM_CUTOFF / 2) / BUTTON_GAP)));
+
+	m_btnBack.m_xPos = m_width - m_btnBack.m_width - 5;
+	m_btnBack.m_yPos = m_height - 27;
+	m_buttons.push_back(&m_btnBack);
+
+	bool flipLeft = false;
+	for (int i = 0; i < KM_COUNT; i++)
+	{
+		if (flipLeft)
+			m_rebindingOptions[i].button->m_xPos = m_width / 2;
+		else
+			m_rebindingOptions[i].button->m_xPos = m_width / 12;
+		
+		m_rebindingOptions[i].button->m_yPos = m_rebindingOptions->ypos;
+		flipLeft = !flipLeft;
+	}
+
+	for (int i = 0; i < KM_COUNT; i++)
+		m_buttons.push_back(m_rebindingOptions[i].button);
+
+	for (int i = 0; i < KM_COUNT; i++)
+		m_buttonTabList.push_back(m_rebindingOptions[i].button);
+		
+	GetRenderableButtons();
+	UpdateText();
+}
+
+void ControlsScreen::render(int a, int b, float c)
+{
+	if (!m_pMinecraft->isLevelGenerated())
+		renderMenuBackground(c);
+
+	fillGradient(0, TOP_CUTOFF - 2, m_width, m_height - BOTTOM_CUTOFF + 7, 0xC0101010, 0xD0101010);
+	//Won't work?
+	//fill(0, 0, m_width, TOP_CUTOFF, 0xFFFFFF);
```

Comment:
Don't know if this was intentional because of [here](https://discord.com/channels/1135450987096703096/1135450987549704234/1143715947870433300) but I would recommend using `renderBackground()` when we're in a level.
```cpp
if (!m_pMinecraft->isLevelGenerated())
{
	renderMenuBackground(c);
	fillGradient(0, TOP_CUTOFF - 2, m_width, m_height - BOTTOM_CUTOFF + 7, 0xC0101010, 0xD0101010);
}
else
{
	renderBackground();
}
```

Result:
![UIhcgFAfOk](https://github.com/ReMinecraftPE/mcpe/assets/124226059/5a489a1d-cb15-4c55-b572-7a5736be40fd)